File: backend-spring/src/main/java/org/lafeuille/demo/services/PublicHolidaysService.java
package org.lafeuille.demo.services;

import static java.util.stream.Collectors.toList;

import de.focus_shift.jollyday.core.Holiday;
import de.focus_shift.jollyday.core.HolidayManager;
import de.focus_shift.jollyday.core.ManagerParameters;
import java.time.Year;
import java.util.Comparator;
import java.util.List;
import java.util.Locale;
import java.util.Optional;
import org.springframework.stereotype.Service;
import org.threeten.extra.LocalDateRange;

@Service
public class PublicHolidaysService {

  public List<Holiday> getHolidays(LocalDateRange dateRange, Locale locale, Optional<String> country) {
    return getManager(locale, country)
        .getHolidays(dateRange.getStart(), dateRange.getEnd())
        .stream()
        .sorted(Comparator.comparing(Holiday::getDate))
        .collect(toList());
  }

  public List<Holiday> getYearHolidays(Year isoYear, Locale locale, Optional<String> country) {
    return getManager(locale, country)
        .getHolidays(isoYear)
        .stream()
        .sorted(Comparator.comparing(Holiday::getDate))
        .collect(toList());
  }

  private HolidayManager getManager(Locale locale, Optional<String> country) {
    var managerParameter = country.map(ManagerParameters::create)
        .orElseGet(() -> ManagerParameters.create(locale));
    return HolidayManager.getInstance(managerParameter);
  }

}

File: backend-spring/src/main/java/org/lafeuille/demo/services/PublicHolidaysService.kt
package org.lafeuille.demo.services

import de.focus_shift.jollyday.core.Holiday
import de.focus_shift.jollyday.core.HolidayManager
import de.focus_shift.jollyday.core.ManagerParameters
import org.springframework.stereotype.Service
import org.threeten.extra.LocalDateRange
import java.time.Year
import java.util.Locale
import java.util.Optional
import java.util.function.Function
import java.util.function.Supplier
import java.util.stream.Collectors

@Service
class PublicHolidaysService {
    fun getHolidays(
        dateRange: LocalDateRange,
        locale: Locale,
        calendar: Optional<String>,
    ): List<Holiday> =
        getManager(locale, calendar)
            .getHolidays(dateRange.start, dateRange.end)
            .stream()
            .sorted(Comparator.comparing(Function { obj: Holiday -> obj.date }))
            .collect(Collectors.toList())

    fun getYearHolidays(
        isoYear: Year,
        locale: Locale,
        calendar: Optional<String>,
    ): List<Holiday> =
        getManager(locale, calendar)
            .getHolidays(isoYear)
            .stream()
            .sorted(Comparator.comparing(Function { obj: Holiday -> obj.date }))
            .collect(Collectors.toList())

    private fun getManager(
        locale: Locale,
        calendar: Optional<String>,
    ): HolidayManager {
        val managerParameter =
            calendar
                .map(Function { calendarPart: String -> ManagerParameters.create(calendarPart) })
                .orElseGet(Supplier { ManagerParameters.create(locale) })
        return HolidayManager.getInstance(managerParameter)
    }
}
